### ISSUE TITLE
Misuse of Math.random() to generate random integers

### ISSUE THREAD

**Issue by dimitri-xyz** [brave/browser-laptop#6944]
In     
https://github.com/brave/browser-laptop/blob/57e69652217e69c937407f1f776b370f75dbd38f/app/extensions/brave/content/scripts/adInsertion.js#L74-L75

That is:
```javascript
var segments = ['IAB2', 'IAB17', 'IAB14', 'IAB21', 'IAB20']
var segment = segments[Math.floor(Math.random() * 4)]
```
The last value 'IAB20' is *excluded* from the range of random() and will never be chosen.

Here's another bug:
https://github.com/brave/browser-laptop/blob/57e69652217e69c937407f1f776b370f75dbd38f/tools/lib/transactionHelpers.js#L48

```javascript
 const count = Math.round(Math.random() * 100)
```
In this case, both 0 and 100 are part of the results, but only occur half as frequently as the other numbers. To see why this happens consider:

```javascript
 const zeroOneTwo = Math.round(Math.random() * 2)
```
all numbers from 0 to 0.49999 will return 0
all numbers from 0.5 to 1.4999 will return 1
all numbers from 1.5 to 1.9999 will return 2

So 1 occurs with probbility 1/2 and 0 and 2 are returned with probability 1/4 each.
This is very different from the expected uniform distribution (1/3,1/3,1/3).

The problem here is the flooring or rounding of floating point random numbers to obtain integers.
I think the "proper" way to obtain random integers in a given range is to use `randomInt` from `random-lib`

*There are many more instances of this in the code. Just do a search for random() to find them.*

**Comment by bsclifton** [comment]
We do use random-lib in our ledger code; it wouldn't be hard to rework instances to call that instead

[Docs on Math.random() explaining the exclusivity issue](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random) shown in the first example (4 never being possible because 1 is excluded from Math.random results)

**Comment by Liunkae** [comment]
Which instances are worth fixing? https://github.com/brave/browser-laptop/search?l=JavaScript&q=random&type=Code&utf8=%E2%9C%93
Here is an example for the new tab page backgrounds: ([js/about/newtab.js](https://github.com/brave/browser-laptop/blob/master/js/about/newtab.js))
````js
@@ -19,7 +19,7 @@ const urlutils = require('../lib/urlutil')
 const siteTags = require('../constants/siteTags')
 const config = require('../constants/config')
 const backgrounds = require('../data/backgrounds')
-const {random} = require('../../app/common/lib/randomUtil')
+const random = require('random-lib')
 
 const ipc = window.chrome.ipcRenderer
 
@@ -65,7 +65,7 @@ class NewTabPage extends React.Component {
     return this.state.showImages && !!this.state.backgroundImage
   }
   get randomBackgroundImage () {
-    const image = Object.assign({}, backgrounds[Math.floor(random() * backgrounds.length)])
+    const image = Object.assign({}, backgrounds[random.randomInt({ min: 0, max: backgrounds.length })])
     image.style = {backgroundImage: 'url(' + image.source + ')'}
     return image
   }
````

**Comment by dimitri-xyz** [comment]
@Liunkae ,

Fixing this turns out to be more complicated than I thought. There are 2 separate issues here:

1. Inadvertently excluding boundary values when generating random numbers
2. Generating a non-uniform (i.e. biased) probability distribution

I think we should fix (1) now, but leave (2) for later.

To fix the first issue, fix all instances where a floating-point random value obtained through `random()` is later turned into an Integer. The conversion can be made by truncation, rounding, flooring, etc. The key is to think *floating-point random number later converted into an integer*. We just have to be careful to know whether the "large" boundary value should be included. [Your example](https://github.com/brave/browser-laptop/issues/6944#issuecomment-287518751) looks good, but I don't know the code and, thus I don't know if we should include only `length-1` or should also include `length`. You'd have to investigate each case. Maybe @bsclifton or someone else should pitch in here on how to do that.

The second problem (non-uniform distribution) turns out to be worse than I thought. The [Mozilla Docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random) are suggesting an *incorrect* way to do it (`getRandomInt` and `getRandomIntInclusive` are biased) and random-lib's `randomInt()` is also incorrectly generating its integers and causing a small bias. The bias is *really small* in most cases, so for most applications it will be fine. I will work with them to fix those issues. 

The correct way to do generate random integers is [Rejection Sampling](https://en.wikipedia.org/wiki/Rejection_sampling).

**Comment by fardog** [comment]
_I'm the author of `random-lib`._

@Liunkae for the options passed to `randomInt`, `min` is inclusive and `max` is exclusive, so that would be the correct fix. (see [docs](https://github.com/fardog/node-random-lib#options))

As to fixing the bias issue, @dimitri-xyz has written up a [very thorough blog](http://dimitri.xyz/random-ints-from-random-bits/) post detailing the issue with the [algorithm mozilla details](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random) and i'll be attempting to fix this issue over the next week or so (if anyone has the time to get to this faster, pull requests are very welcome!). I'll update here once a version is available that removes the bias.

**Comment by 333aleix333** [comment]
Hi! I think I have corrected all the flooring and all the rounding, but I'm new and I don't know what should I do now. 
Thanks!

**Comment by diracdeltas** [comment]
@dimitri-xyz thanks for pointing out this issue and the detailed explanations. Although we are not currently using Math.random and random-lib for purposes that require cryptographic randomness (such as generating crypto keys), I labeled this issue under security to make sure that these libraries are not misused in the future.

@333aleix333 thanks. please fork our repository, push your changes to the fork, then open a pull request against our repository. see here for some instructions: https://gist.github.com/Chaser324/ce0505fbed06b947d962

**Comment by fardog** [comment]
random-lib has been updated to version 3.0, which should solve this issue; see [here](https://github.com/fardog/node-random-lib/blob/master/CHANGELOG.md#300) for the changes. It's now backed by [rejection-sampled-int](https://github.com/fardog/rejection-sampled-int/) for integers, which I also wrote. I would definitely appreciate additional sets of eyes on it, as it's not gone through any formal audit, yet I have tested the features in both libraries heavily.

The API also changed in this version as the old one had behaviors that were very awkward, however conversion to the new API is super-straightforward since there wasn't much surface there to begin with.

Let me know if you have any questions!

**Comment by arsalankhalid** [comment]
If that's the case, you guys should close this issue! @bsclifton

**Comment by fardog** [comment]
There were API changes between v2 and v3. It should not be very involved, but it will require a few more changes. See https://github.com/fardog/node-random-lib/blob/master/CHANGELOG.md#300

I had intended to submit a pr, but wasn't able to get the brave browser tests working on my local machine in the short time I was able to allocate. If someone has the chance though, it should not be a significant change set.
